okhttp/src/main/java/okhttp3/ConnectionSpec.kt

```diff
@@ -0,0 +1,324 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3
+
+import okhttp3.ConnectionSpec.Builder
+import okhttp3.internal.Util
+import okhttp3.internal.Util.concat
+import okhttp3.internal.Util.indexOf
+import okhttp3.internal.Util.intersect
+import okhttp3.internal.Util.nonEmptyIntersection
+import okhttp3.internal.Util.tlsVersionsForJavaNames
+import java.util.Arrays
+import java.util.Objects
+import javax.net.ssl.SSLSocket
+
+/**
+ * Specifies configuration for the socket connection that HTTP traffic travels through. For `https:`
+ * URLs, this includes the TLS version and cipher suites to use when negotiating a secure
+ * connection.
+ *
+ * The TLS versions configured in a connection spec are only be used if they are also enabled in the
+ * SSL socket. For example, if an SSL socket does not have TLS 1.3 enabled, it will not be used even
+ * if it is present on the connection spec. The same policy also applies to cipher suites.
+ *
+ * Use [Builder.allEnabledTlsVersions] and [Builder.allEnabledCipherSuites] to defer all feature
+ * selection to the underlying SSL socket.
+ *
+ * The configuration of each spec changes with each OkHttp release. This is annoying: upgrading
+ * your OkHttp library can break connectivity to certain web servers! But it’s a necessary annoyance
+ * because the TLS ecosystem is dynamic and staying up to date is necessary to stay secure. See
+ * [OkHttp's TLS
+ * Configuration History](https://github.com/square/okhttp/wiki/TLS-Configuration-History) to track
+ * these changes.
+ */
+class ConnectionSpec internal constructor(builder: Builder) {
+  val isTls: Boolean = builder.tls
+  internal val supportsTlsExtensions: Boolean = builder.supportsTlsExtensions
+  internal val cipherSuites: Array<String>? = builder.cipherSuites
+  internal val tlsVersions: Array<String>? = builder.tlsVersions
+
+  /**
+   * Returns the cipher suites to use for a connection. Returns null if all of the SSL socket's
+   * enabled cipher suites should be used.
+   */
+  fun cipherSuites(): List<CipherSuite>? {
+    return if (cipherSuites != null) CipherSuite.forJavaNames(*cipherSuites) else null
+  }
+
+  /**
+   * Returns the TLS versions to use when negotiating a connection. Returns null if all of the SSL
+   * socket's enabled TLS versions should be used.
+   */
+  fun tlsVersions(): List<TlsVersion>? {
+    return if (tlsVersions != null) tlsVersionsForJavaNames(*tlsVersions) else null
```

Comment:
I think you don't like it but `return tlsVersions?.let { tlsVersionsForJavaNames(*it) }`

Reply:
We shouldn't need any helpers that convert an array/list to a list with a transformation anymore. This could be `tlsVersions?.map(::tlsVersionForJavaName)`, right?

Reply:
Oooh, nice. Fixed.